### PR TITLE
implement scan and skip rules

### PR DESCRIPTION
**Pull request by patilpankaj212**
1. command line options to scan and skip specific rules
2. config file based rule scanning and skipping
3. unit tests

### CODE REVIEW

File: pkg/http-server/file-scan.go [hunk 1]
```diff
@@ -86,10 +86,10 @@ func (g *APIHandler) scanFile(w http.ResponseWriter, r *http.Request) {
 	var executor *runtime.Executor
 	if g.test {
 		executor, err = runtime.NewExecutor(iacType, iacVersion, cloudType,
-			tempFile.Name(), "", "", []string{"./testdata/testpolicies"})
+			tempFile.Name(), "", "", []string{"./testdata/testpolicies"}, []string{}, []string{})
```
Comment (kanchwala-yusuf) [inline pkg/http-server/file-scan.go:89]: We should be taking rules to scan and skip for the API server as well. This functionality applies to the scanning for API server as well.

File: pkg/http-server/remote-repo.go [hunk 2]
```diff
@@ -112,7 +112,7 @@ func (s *scanRemoteRepoReq) ScanRemoteRepo(iacType, iacVersion string, cloudType
 
 	// create a new runtime executor for scanning the remote repo
 	executor, err := runtime.NewExecutor(iacType, iacVersion, cloudType,
-		"", iacDirPath, "", policyPath)
+		"", iacDirPath, "", policyPath, []string{}, []string{})
```
Comment (kanchwala-yusuf) [inline pkg/http-server/remote-repo.go:115]: We should be taking rules to scan and skip for the API server as well. This functionality applies to the scanning for API server as well.

File: pkg/policy/opa/engine.go [hunk 3]
```diff
@@ -46,13 +46,13 @@ var (
 )
 
 // NewEngine returns a new OPA policy engine
-func NewEngine(policyPath string) (*Engine, error) {
+func NewEngine(policyPath string, scanRules, skipRules []string) (*Engine, error) {
```
Comment (kanchwala-yusuf) [inline pkg/policy/opa/engine.go:49]: Wondering if `scanRules` and `skipRules` can become part of the `Engine{}` struct. 

Would like an opinion from @williepaul on this.

File: pkg/policy/opa/engine.go [hunk 4]
```diff
@@ -249,14 +249,20 @@ func (e *Engine) CompileRegoFiles() error {
...
+func (e *Engine) Init(policyPath string, scanRules, skipRules []string) error {
 	e.context = context.Background()
 
 	if err := e.LoadRegoFiles(policyPath); err != nil {
 		zap.S().Error("error loading rego files", zap.String("policy path", policyPath), zap.Error(err))
 		return err
 	}
 
+	// before compiling the rego files, filter the rules based on scan and skip rules supplied
+	filterRules(e, policyPath, scanRules, skipRules)
```
Comment (kanchwala-yusuf) [inline pkg/policy/opa/engine.go:261]: Would it make sense to have `filterRules` to be tied to the `Engine` struct?

File: pkg/policy/opa/engine.go [hunk 5]
```diff
@@ -249,14 +249,20 @@ func (e *Engine) CompileRegoFiles() error {
...
+
+	// update the rule count
+	e.stats.ruleCount = len(e.regoDataMap)
```
Comment (kanchwala-yusuf) [inline pkg/policy/opa/engine.go:264]: May be it's a good idea to error out if the `ruleCount` is less than 1.

File: pkg/policy/opa/engine.go [hunk 6]
```diff
@@ -395,3 +401,49 @@ func (e *Engine) Evaluate(engineInput policy.EngineInput) (policy.EngineOutput,
...
+		filterScanRules(e, policyPath, scanRules)
+	}
+
+	// apply skip rules
+	if len(skipRules) > 0 {
+		filterSkipRules(e, policyPath, skipRules)
+	}
+}
+
+func filterScanRules(e *Engine, policyPath string, scanRules []string) {
```
Comment (kanchwala-yusuf) [inline pkg/policy/opa/engine.go:417]: Can we tie `filterScanRules` with the `Engine{}` struct?

File: pkg/policy/opa/engine.go [hunk 7]
```diff
@@ -395,3 +401,49 @@ func (e *Engine) Evaluate(engineInput policy.EngineInput) (policy.EngineOutput,
...
+	}
+	if len(tempMap) == 0 {
+		zap.S().Warnf("scan rule id's: %+v not found in policy path: %s", scanRules, policyPath)
+	}
+
+	// the regoDataMap should only contain regoData for supplied scan rules
+	e.regoDataMap = tempMap
+}
+
+func filterSkipRules(e *Engine, policyPath string, skipRules []string) {
```
Comment (kanchwala-yusuf) [inline pkg/policy/opa/engine.go:438]: Can we tie `filterSkipRules` with the `Engine{}` struct?

File: pkg/policy/opa/engine_test.go [hunk 8]
```diff
@@ -1 +1,98 @@
...
+					regoDataMap: getTestRegoDataMap(10),
+				},
+				policyPath: testPolicyPath,
+				scanRules:  []string{"Rule.11", "Rule.12", "Rule.13"},
+			},
+			assert:      true,
+			regoMapSize: 0,
+		},
+		{
+			name: "skip rules test",
```
Comment (kanchwala-yusuf) [inline pkg/policy/opa/engine_test.go:54]: Can we have a test for `scan and skip rules present`?

File: pkg/runtime/executor.go [hunk 9]
```diff
@@ -17,12 +17,27 @@
...
 
+	"github.com/accurics/terrascan/pkg/config"
 	iacProvider "github.com/accurics/terrascan/pkg/iac-providers"
 	"github.com/accurics/terrascan/pkg/notifications"
 	"github.com/accurics/terrascan/pkg/policy"
 	opa "github.com/accurics/terrascan/pkg/policy/opa"
+	"github.com/pelletier/go-toml"
+)
+
+const (
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:32]: I would suggest creating a new file `rules.go` and move these constants in that file.

File: pkg/runtime/executor.go [hunk 10]
```diff
@@ -17,12 +17,27 @@
...
+	rulesKey     = "rules"
+	scanRulesKey = "scan-rules"
+	skipRulesKey = "skip-rules"
+)
+
+var (
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:38]: Another candidate for `rules.go`

File: pkg/runtime/executor.go [hunk 11]
```diff
@@ -139,3 +168,68 @@ func (e *Executor) Execute() (results Output, err error) {
 	// successful
 	return results, nil
 }
+
+// read the config file and update scan and skip rules
+func (e *Executor) initScanAndSkipRules() error {
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:173]: Another good candidate for `rules.go`

File: pkg/runtime/executor.go [hunk 12]
```diff
@@ -139,3 +168,68 @@ func (e *Executor) Execute() (results Output, err error) {
...
+			// read skip rules in the toml tree
+			if err := initRules(e, data, skipRulesKey); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func initRules(e *Executor, tree *toml.Tree, key string) error {
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:199]: Another good candidate for `rules.go`

File: pkg/runtime/executor.go [hunk 13]
```diff
@@ -139,3 +168,68 @@ func (e *Executor) Execute() (results Output, err error) {
...
+		} else {
+			e.skipRules = append(e.skipRules, rules...)
+		}
+	} else {
+		zap.S().Debugf("key '%s' not found in the config file: %s", key, e.configFile)
+	}
+	return nil
+}
+
+func getRulesInTomlTree(tree *toml.Tree, configFile, key string) ([]string, error) {
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:217]: Another good candidate for `rules.go`.

But, I feel this should function should be a part of the `config` package. This function can be a helper for returning rules from the `config` package.

File: pkg/runtime/executor.go [hunk 14]
```diff
@@ -139,3 +168,68 @@ func (e *Executor) Execute() (results Output, err error) {
...
+
+func initRules(e *Executor, tree *toml.Tree, key string) error {
+	rules, err := getRulesInTomlTree(tree, e.configFile, key)
+	if err != nil {
+		zap.S().Error("error reading config file", zap.Error(err))
+		return err
+	}
+	if len(rules) > 0 {
+		if key == scanRulesKey {
+			e.scanRules = append(e.scanRules, rules...)
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:207]: Not the right place to append to scan rules, this function should just return the list of rules. Caller function should take responsibility of using the list of rules. In this way this function can become more generic and we would not require to pass the `Executor` object to this.

File: pkg/policy/interface.go [hunk 15]
```diff
@@ -18,7 +18,8 @@ package policy
 
 // Engine Policy Engine interface
 type Engine interface {
-	Init(string) error
+	// to initialize engine with policy path, scan and skip rules
```
Comment (kanchwala-yusuf) [inline pkg/policy/interface.go:21]: Comment should start with `Init`

File: pkg/runtime/executor.go (hunk 11 above)
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:173]: Also, `initRules()` as a function name should suffice, isn't it?

File: pkg/runtime/executor.go [hunk 16]
```diff
@@ -139,3 +168,68 @@ func (e *Executor) Execute() (results Output, err error) {
 	// successful
 	return results, nil
 }
+
+// read the config file and update scan and skip rules
+func (e *Executor) initScanAndSkipRules() error {
+	if e.configFile != "" {
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:174]: Would be a better idea to do:
```
if e.configFile == "" {
    return nil
}
```

File: pkg/runtime/executor.go [hunk 17]
```diff
@@ -139,3 +168,68 @@ func (e *Executor) Execute() (results Output, err error) {
...
+		configData, err := config.LoadConfig(e.configFile)
```
Comment (kanchwala-yusuf) [inline pkg/runtime/executor.go:175]: It should not be the responsibility of this package to `LoadConfig`, it should happen in `config` package. There should be handlers in the `config` package to return what's required by this package.

**Review by kanchwala-yusuf — CHANGES_REQUESTED** [review]

File: pkg/http-server/file-scan.go (hunk 1 above)
Comment (patilpankaj212) [reply pkg/http-server/file-scan.go]: Implemented

File: pkg/http-server/remote-repo.go (hunk 2 above)
Comment (patilpankaj212) [reply pkg/http-server/remote-repo.go]: Implemented

File: pkg/policy/opa/engine.go (hunk 3 above)
Comment (patilpankaj212) [reply pkg/policy/opa/engine.go]: Added a function FilterRules(string, []string, []string) to Engine interface and moved the Init function out of the NewEngine function.

File: pkg/policy/opa/engine.go (hunk 4 above)
Comment (patilpankaj212) [reply pkg/policy/opa/engine.go]: implemented as part of interface

File: pkg/policy/opa/engine.go (hunk 5 above)
Comment (patilpankaj212) [reply pkg/policy/opa/engine.go:258]: multiple engines are created based on policy path slice, cannot error out.

File: pkg/policy/opa/engine_test.go (hunk 8 above)
Comment (patilpankaj212) [reply pkg/policy/opa/engine_test.go:49]: added a test for it.

File: pkg/runtime/executor.go (hunk 13 above)
Comment (patilpankaj212) [reply pkg/runtime/executor.go]: this was removed as part of config reader refactor

File: pkg/runtime/executor.go (hunk 14 above)
Comment (patilpankaj212) [reply pkg/runtime/executor.go]: removed in the refactored code

File: pkg/runtime/executor.go (hunk 17 above)
Comment (patilpankaj212) [reply pkg/runtime/executor.go]: done as part of config reader refactor

File: pkg/policy/opa/engine.go (hunk 4 above)
Comment (williepaul) [reply pkg/policy/opa/engine.go]: At this point, adding to an interface makes sense. Later, we should have a top-level "Scanner" struct that manages interactions between the scan engine, filters, policies/rules, and results. Today, that's all considered the "scan engine", but I just want to limit that piece to actual scanning only. Policies, rules, rule filters, and results processing are applicable to different scan engines, so in the future they should work independently of it.

File: pkg/config/global.go [hunk 18]
```diff
@@ -31,87 +30,54 @@ const (
...
+	configReader, err := NewTerrascanConfigReader(configFile)
 	if err != nil {
-		return p, ErrNotPresent
+		return err
 	}
 
-	keyConfig := config.Get(policyConfigKey)
-	if keyConfig == nil {
-		return p, errTomlKeyNotPresent
+	if len(configReader.GetPolicyConfig().BasePath) > 0 {
```
Comment (williepaul) [inline pkg/config/global.go:68]: I think we shouldn't do these length checks. The reason is, if the global config happens to contain an existing config, running loadConfigFile could leave the user with a partially loaded config file (values from old config + some values from new config), and that is misleading to the user. I would expect if the new config file has an empty path, an empty path should be loaded. That is an error the user would have to fix. I would just remove the length check and set it directly: Global.Policy.BasePath = configReader.GetPolicyConfig().BasePath, or simply assign the entire config in one shot.

**Review by williepaul — CHANGES_REQUESTED** [review]

**Review by kanchwala-yusuf — APPROVED** [review]

**Review by williepaul — APPROVED** [review]

**Pull request merged**